sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnReflector.java

```diff
@@ -91,1012 +36,268 @@
-
-      // First, find all declared methods on the startClass and parents (up to stopClass)
-      while (clazz != null && !clazz.equals(stopClass)) {
-        for (Method method : clazz.getDeclaredMethods()) {
-          if (method.isAnnotationPresent(anno)) {
-            matches.add(method);
-          }
+      if (param.isSubtypeOf(BoundedWindow.class)) {
+        if (extraParameters.contains(DoFnSignature.ProcessElementMethod.Parameter.BOUNDED_WINDOW)) {
+          throw new IllegalArgumentException("Multiple BoundedWindow parameters");
         }
-
-        Collections.addAll(interfaces, clazz.getInterfaces());
-
-        clazz = clazz.getSuperclass();
-      }
-
-      // Now, iterate over all the discovered interfaces
-      for (Method method : ReflectHelpers.getClosureOfMethodsOnInterfaces(interfaces)) {
-        if (method.isAnnotationPresent(anno)) {
-          matches.add(method);
+        extraParameters.add(DoFnSignature.ProcessElementMethod.Parameter.BOUNDED_WINDOW);
+      } else if (param.isSubtypeOf(DoFn.InputProvider.class)) {
+        if (extraParameters.contains(DoFnSignature.ProcessElementMethod.Parameter.INPUT_PROVIDER)) {
+          throw new IllegalArgumentException("Multiple InputProvider parameters");
         }
-      }
-      return matches;
-    }
-
-    private static Method findAnnotatedMethod(
-        Class<? extends Annotation> anno, Class<?> fnClazz, boolean required) {
-      Collection<Method> matches = declaredMethodsWithAnnotation(
-          anno, fnClazz, DoFn.class);
-
-      if (matches.size() == 0) {
-        if (required) {
-          throw new IllegalStateException(String.format(
-              "No method annotated with @%s found in %s",
-              anno.getSimpleName(), fnClazz.getName()));
-        } else {
-          return null;
+        if (!param.isSupertypeOf(expectedInputProviderT)) {
+          throw new IllegalArgumentException(
+              String.format(
+                  "Wrong type of InputProvider parameter for method %s: %s, should be %s",
+                  format(m), formatType(param), formatType(expectedInputProviderT)));
         }
-      }
-
-      // If we have at least one match, then either it should be the only match
-      // or it should be an extension of the other matches (which came from parent
-      // classes).
-      Method first = matches.iterator().next();
-      for (Method other : matches) {
-        if (!first.getName().equals(other.getName())
-            || !Arrays.equals(first.getParameterTypes(), other.getParameterTypes())) {
-          throw new IllegalStateException(String.format(
-              "Found multiple methods annotated with @%s. [%s] and [%s]",
-              anno.getSimpleName(), format(first), format(other)));
+        extraParameters.add(DoFnSignature.ProcessElementMethod.Parameter.INPUT_PROVIDER);
+      } else if (param.isSubtypeOf(DoFn.OutputReceiver.class)) {
+        if (extraParameters.contains(
```

Comment:
I added formatting of the current parameter, but storing the type of the previously found parameter seems like it would complicate the code for not very much gain. I doubt that somebody will have trouble finding where is their second OutputReceiver parameter in the method.